[APPSEC-58221] AAP for Go SDK (#30305)
DataDog/documentation: content/en/security/application_security/setup/go/sdk.md

---
title: App and API Protection SDK for Go
further_reading:
- link: "https://pkg.go.dev/github.com/DataDog/dd-trace-go/v2/appsec"
  tag: "Documentation"
  text: "Go Security API docs"
- link: "https://github.com/DataDog/appsec-go-test-app"
  tag: "Source Code"
  text: "Sample application used for exhaustive SDK testing"
- link: "/security/application_security/add-user-info/"
  tag: "Documentation"
  text: "Adding user information to traces"
- link: "/security/application_security/troubleshooting"
  tag: "Documentation"
  text: "Troubleshooting App and API Protection"
---

> Sometimes, instrumenting your application automatically using [orchestrion][12] is not enough.
> That is why Datadog offer an API under [github.com/DataDog/dd-trace-go/v2/appsec][2] for more flexibility and more features
> that would not always be available.

Reviewer: ```suggestion
If you need flexibility and features beyond those available when instrumenting your application automatically using [Orchestrion][12], Datadog provides an App and API Protection API located at [github.com/DataDog/dd-trace-go/v2/appsec][2]. This API improves flexibility and offers additional features.
```

> ## Preamble: errors
> 
> The API is designed to return errors when something goes wrong, so you can handle them in your code. More specifically
> an error of type [`events.BlockingSecurityEvent`][3] is returned when a blocking event occurs, such as a user, IP, or a request being blocked.
> 
> **It is imperative to not write any response to the client when you receive a `BlockingSecurityEvent` error**, otherwise you are
> potentially sending sensitive information to an attacker. A helper called [`events.IsSecurityError`][4] is here to help
> this process. App and API Protection SDK is designed to send a blocking response in your stead (by default, a HTTP 403) when a
> `BlockingSecurityEvent` is returned. Customizing this behavior is possible, more available in
> the [Customize response to blocked requests][5] section.

Reviewer: ```suggestion
## Error event handling

The API error type [`events.BlockingSecurityEvent`][3] is returned when a blocking event occurs, such blocking a user, IP, or request.

**Do not write any response to the client when you receive a `BlockingSecurityEvent` error**. A response could potentially send sensitive information to an attacker. To safely detect this type of error, use the helper function [events.IsSecurityError][4].

When a `BlockingSecurityEvent` is returned, the SDK automatically sends a blocking response for you, typically an HTTP 403 Forbidden.

Customizing this behavior is possible. See the [Customize response to blocked requests][5] section.
```

## HTTP body monitoring

> By default, a lot of Go HTTP Frameworks are not high level enough to provide parsed HTTP request & response bodies.
> This is why App and API Protection SDK provides a way to monitor the HTTP request and response bodies manually:

Reviewer: ```suggestion
By default, a lot of Go HTTP Frameworks are not high level enough to provide parsed HTTP request and response bodies. Consequently, the App and API Protection SDK provides a way to monitor the HTTP request and response bodies manually:
```

- [`appsec.MonitorParsedHTTPBody`][8] to monitor the parsed request body.
- [`appsec.MonitorHTTPResponseBody`][9] to monitor the response body.

> These functions automatically block the current request if asked from Datadog's UI, and return a

Reviewer: ```suggestion
These functions automatically block the current request if App and API Protection is enabled in the Datadog UI, and return a
```

Reviewer: I wasn't sure what you meant by `if asked from Datadog's UI` but I believe calls to this function are ignored if
AAP is disabled or the given context is incorrect, so I suggested `if App and API Protection is enabled in the Datadog UI`.

`BlockingSecurityEvent` error.

## User tracking

> When a user logs in to your application, it is important to track this event for security purposes.
> App and API Protection offers a feature to [track and block login requests][1]. This feature is available under 3 functions:
> 
> - [`appsec.TrackUserLoginSuccess`][10] to track a successful login event on your login endpoint. This function also calls [`appsec.SetUser`][7] internally to associate the user with the current request.
> - [`appsec.TrackUserLoginFailure`][6] to track a failed login event on your login endpoint.
> - [`appsec.SetUser`][7] use on all authenticated requests to report users to Datadog and enable user blocking.

Reviewer: ```suggestion
User login events can be tracked and blocked using the [track and block login requests][1] feature from these functions:

- [`appsec.TrackUserLoginSuccess`][10] tracks a successful login event on your login endpoint. This function also calls [`appsec.SetUser`][7] internally to associate the user with the current request.
- [`appsec.TrackUserLoginFailure`][6] tracks a failed login event on your login endpoint.
- [`appsec.SetUser`][7] can be used on all authenticated requests to report users to Datadog and enable user blocking.
```

> These functions automatically block the user if asked from Datadog's UI, and return a `BlockingSecurityEvent`

Reviewer: I'm not sure what `if asked from Datadog's UI` means. Do you mean enabled in the UI or something?

Author: ```suggestion
These functions automatically block the current request if the [Protection][14] feature is enabled, and return a `BlockingSecurityEvent`
```

error.

## Custom event tracking

> User login success and failure are what is called [Business logic events][11] behind the scene which are events that
> are not directly related to security, but are important for security purposes. Datadog offer the capability to track custom
> events using the [`appsec.TrackCustomEvent`][13] function. This function sets custom event data as service entry span tags,
> which triggers backend-side event monitoring that can ultimately block IP addresses and/or user IDs associated with them.

Reviewer: ```suggestion
You can track [Business logic events][11] like user login success/failure using the [`appsec.TrackCustomEvent`][13] function. This function sets custom event data as service entry span tags, which triggers backend event monitoring, and can ultimately block IP addresses and/or associated user IDs.
```

> ## Real world example

Reviewer: ```suggestion
## Example
```
